### ISSUE TITLE
Bad diagnostic when pattern matching `nil` that cannot be inferred

### ISSUE THREAD

**Issue by hamishknight** [apple/swift#64279]
The following gives a bad diagnostic:

```swift
func foo<T>() -> T { fatalError() }
if nil ~= foo() {} // error: Expression pattern of type 'String' cannot match values of type 'Substring'
```

**Comment by Rajveer100** [comment]
@hamishknight If I correctly understand the issue, the function `foo<T>() -> T` is a generic function with type `T` and throws a `fatalError()`.

Next, the `if` block checks if `nil` contains the function `foo()` (which I am unsure of, but am aware of the operator `~=` used as range). 

Let me know to what I am missing out, and some brief description to get an idea!

**Comment by AnthonyLatsis** [comment]
> Next, the `if` block checks if `nil` contains the function `foo()`

Not quite. See [Expression Pattern](https://docs.swift.org/swift-book/documentation/the-swift-programming-language/patterns/#Expression-Pattern).

**Comment by Rajveer100** [comment]
Does it check if the return value of the function is 'nil'?

**Comment by Rajveer100** [comment]
Ok, I changed the `snippet code` this way, now it does throw `fatalError()`:

```
import UIKit

func foo<T>() -> T { fatalError() }
if "nil" ~= foo() {} // error: Fatal Error

```

**Comment by hamishknight** [comment]
Indeed, but the `nil` case is invalid because there's no type to infer for it, and we should ideally diagnose as such. What's currently happening is that the constraint system is trying each overload of `~=`, and coming up with a bunch of bad solutions, and it just so happens that the `func ~= (lhs: Substring, rhs: String) -> Bool` solution is the least bad one, and we attempt to diagnose based on that.

**Comment by Rajveer100** [comment]
I see!

**Comment by Rajveer100** [comment]
After some finding, I managed to find the call point in `ConstraintSystem.h` under this `if` condition:

```
else if (operatorName == ctx.Id_MatchOperator) {
      DE.diagnose(anchor->getLoc(), diag::cannot_match_expr_pattern_with_value,
                  lhsType, rhsType);
    }
```

So we could probably add an `check` inside this to first check for the `nil` case as in the issue and add a more specific message?

**Comment by Rajveer100** [comment]
@hamishknight What do you suggest?
